### ISSUE TITLE
No modified base tags for reads with supplementary alignments?

### ISSUE THREAD

**Issue by PanZiwei** [epi2me-labs/modbam2bed#47]
Hi I am specifically interested in the reads with supplementary alignments, but I failed to extract the base modified tags and flag with the `modbampy` library. 
The code below can work on a regular read with primary alignments - the `print(*pos_mod)` works well. But for the read with supplementary alignments, the output is blank.
I checked with samtools and found that it seems that the ML tag DOESN'T exist for the reads with supplementary alignments, is it expected?

Any thoughts to solve the problem? Thank you so much for your help!

```
#### Read with supplementary
input_bam='/fastscratch/c-panz/test/test.chr8.supplementary.bam'
for chr_name, chr_size in chr_list:
    with ModBam(input_bam) as bam:
        for read in bam.reads(chr_name, 0, int(chr_size)):
            read_data=[]
            print(read._bam1_t.core.flag)
            for pos_mod in read.mod_sites:
                print(*pos_mod)

Blank output
```
The way I am checking whether there is an ML tag or not in read with supplementary alignments:

```
(modbam2bed) [c-panz@sumner-log1 test]$ samtools view test.chr8.supplementary.bam | awk '{for (i=12; i<=NF; ++i) { if ($i ~ "^MM:"){ split($i, tc, ":"); td[tc[1]] = tc[3]; } }; print "@"$1"\n"td["MM"] }' > test.chr8.supplementary.ML.txt
(modbam2bed) [c-panz@sumner-log1 test]$ vim test.chr8.supplementary.ML.txt

@209a1e4d-ab19-41ab-b108-93f2dae36b21

@dfb14074-5d81-4b86-b4db-41aec80a6718

@209a1e4d-ab19-41ab-b108-93f2dae36b21

@a6aad990-e6a6-45ba-add3-7183f7d5073c

@dfb14074-5d81-4b86-b4db-41aec80a6718

@209a1e4d-ab19-41ab-b108-93f2dae36b21

@a6aad990-e6a6-45ba-add3-7183f7d5073c

@081da8fe-94e3-4f98-ac97-a6e22d49e6e5

**Comment by cjw85** [comment]
The iterator internal to the C library filters all but valid primary alignments. Its not currently possible therefore to get information from supplementary alignments.

https://github.com/epi2me-labs/modbam2bed/blob/master/src/bamiter.c#L134

**Comment by PanZiwei** [comment]
Thanks for the quick reply! I have 2 following-up questions:
1. Is the modified base predicted by Guppy regardless of its alignment status? Or it only predicts the sites that are primary alignments?

2. What if I didn't want the supplementary alignment filter? Can I: copy the repository ->  modify the line you mentioned ->  install it from the modified source code? Any other suggestions to dig out the modified status for supplementary alignment?

Thank you again!

**Comment by cjw85** [comment]
1. The modified base calling in both Guppy and dorado is independent of the reference (or alignment).
2. You can try editing the source. But first I would inspect your BAM file for supplementary alignments and check if those entries have `MM` and `ML` tags. If they do not then `modbam2bed` is not going to be able to extract useful information.

**Comment by PanZiwei** [comment]
Thanks for the tips! I selected the reads that have supplementary alignments only by `samtools view -b -F 4 -f 2048 ${input_bam} > ${output_bam}` and check the `MM` and `ML` tags in the output bams.  However, it's really weird that there is NO entries for `MM` and `ML` at all...

**Comment by cjw85** [comment]
Well, modbam2bed and I can't really help you if there are no tags! 😄 

Additionally, minimap2 (and Guppy) will hard clip supplementary alignments by default, unless run with `-Y`. Hard clipping will invalidate the `MM` and `ML` tags; @jts has provided a PR to htslib that would enable modbam2bed to safely parse alignments with hard clipping (well, ignore them). Until that is merged, modbam2bed will continue to ignore SUP alignments.

**Comment by PanZiwei** [comment]
Thanks for the reply! I am wondering whether I can ask 3 following questions:
1. Can you talk more about the "hard clip supplementary alignments"? What does it mean? 

2. For modbam2bed, does it only check the C in the "CpG" or "GpC" pattern?  I found the sequence is different between the reference genome and the predicted base-calling sequence by Guppy, which sequence do you use when doing the per-site calculation?

3. I redid a Guppy modified base calling for a specific read that contains 1 primary alignment and 1 supplementary alignment and checked the modified information in the bam file with `pysam` package. The results showed that the `MM` and `ML` tags are ONLY detected in the primary alignment, NOT detected in the supplementary alignment. But the sequence are both base-called in the two alignments.

Do you think it is relevant to the "hard clip" parameter? Does the scenario contradict your previous statement, "The modified base calling in both Guppy and dorado is independent of the reference (or alignment)"?  Is there a way to get modified information for supplementary alignment also?

The command I am using to check the tag:
```
with pysam.AlignmentFile(input_bam, "rb") as bam_file:
    # Iterate over the reads in the BAM file
    for read in bam_file:
            print(read.query_name)
            
            try:
                mm_tag = read.get_tag("MM")
                print(len(mm_tag))
            except KeyError:
                print("The read {} DOESN'T have mm tag!".format(read.query_name))
                
            try:
                ml_tag = read.get_tag("ML")
                print(len(ml_tag))
            except KeyError:
                print("The read {} DOESN'T have ml tag!".format(read.query_name))

            # Get the read sequence and count the number of C's
            seq = read.query_sequence
            chrom = bam_file.get_reference_name(read.reference_id)
            pos = read.pos
            print(read.flag, chrom, pos, len(seq))
```

```
### Primary alignment
2a62699a-e18c-4500-b398-7c2e52f0b571
781
291
0 chr1 51040824 38066

### Supplmentary alignment
2a62699a-e18c-4500-b398-7c2e52f0b571
The read 2a62699a-e18c-4500-b398-7c2e52f0b571 DOESN'T have mm tag!
The read 2a62699a-e18c-4500-b398-7c2e52f0b571 DOESN'T have ml tag!
2048 chr14 64722259 5400
```

Thank you so much for your help!

**Comment by cjw85** [comment]
1. hard clipping is something defined in the SAM specification: it is the ability to truncate query sequences in an alignment record.
2. modbam2bed without motif filtering reports all sites with the pre-requisite base. It does this for both strands of the reference. For CpG filtering this means the C on the reference strand and the C on the reverse complemented strand (corresponding to the G on the forward strand). It doesn't check GC, that would be wrong.
3. Consulting with colleagues, Guppy does not attach the modified base information to supplementary alignments.

I don't believe theres any contradictions here, only new information.

> Is there a way to get modified information for supplementary alignment also?

In short no, not without writing code to track from the supplementary alignment to the corresponding primary alignment containing the ML tag and working out the correspondence of the information contained therein to the query sequence of the supplementary alignment.

**Comment by PanZiwei** [comment]
Thanks for the reply! Correct me if I am wrong about the following statements:

1. For the current bam file, the logic is: modified base calling by Guppy and dorado -> Check the alignment type -> If alignment is primary, output the modified base calling into bam file; If not, the info is NOT written into bam file. 

2. Do you have any suggestions to track the supplementary alignment to the corresponding primary alignment?  I did a quick check on the sequence information between primary and supplementary and found they are not the same and the supplementary sequence is not a subset of primary, so I guess I might need to consider other factors to find the relationship.

4. Even though modbam2bed didn't provide motif filtering, it still supports base filtering in later analysis when the user provides the `--chg/--chh/--cpg`?

5. Any idea about why the length of MM and ML tags values are different? And the C site number in the sequence predicted in the base-calling seq in the bam file `seq  = read.query_sequencecount("C")` is different from the C site number in the aligned regions also...

**Comment by cjw85** [comment]
1. Correct
2. Its possible. It won't be implemented in modbam2bed any time soon.
3. modbam2bed provides motif filtering when one of the options `--chg`, `--chh`, or `--cpg` is provided.
4. I'm not sure what you mean, the tags shouldn't be read directly but using the APIs. Their definition can be found in the SAM tags spec if you want to know how to interpret them.
4b. Because basecalling is fallible.

**Comment by PanZiwei** [comment]
I have one more question:
Is the modified base calling information consistent between the primary and supplementary alignments? In other words, the modified base calling information is independent of the reference genome (That's what you stated before).

If that's the case,  can I simply copy the modified base calling information from the primary alignment to the supplementary alignment for the same read? Then I can use modbam2bed for downstream analysis.

**Comment by cjw85** [comment]
The modified base calling in both Guppy and dorado is independent of the reference (or alignment).

You cannot simply copy the modified base information due to the fact that supplementary alignments are typically hard clipped.

Possibly a simple way might be to realign with minimap2 with the option turned on to not hard clip alignments. You will still be left with the fact that modbam2bed currently ignores supplementary alignments.

There is no more help or guidance I can provide here beyond restating what I have already said.

**Comment by PanZiwei** [comment]
Thank you so much for your help! Really helpful!! I will close the issue then.
